core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java

```diff
@@ -422,37 +549,106 @@ public void updateToRemovedBlock(boolean add, long blockId) {
   /**
    * Sets the capacity of the worker in bytes.
    *
+   * You should lock externally with {@link MasterWorkerInfo#lock(EnumSet, boolean)}
+   * with {@link MasterWorkerInfo.LockType#USAGE_LOCK} specified.
+   * An exclusive lock is required.
+   *
    * @param capacityBytesOnTiers used bytes on each storage tier
    */
   public void updateCapacityBytes(Map<String, Long> capacityBytesOnTiers) {
-    mCapacityBytes = 0;
-    mTotalBytesOnTiers = capacityBytesOnTiers;
-    for (long t : mTotalBytesOnTiers.values()) {
-      mCapacityBytes += t;
+    long capacityBytes = 0;
+    mUsage.mTotalBytesOnTiers = capacityBytesOnTiers;
+    for (long t : mUsage.mTotalBytesOnTiers.values()) {
+      capacityBytes += t;
     }
+    mUsage.mCapacityBytes = capacityBytes;
   }
 
   /**
    * Sets the used space of the worker in bytes.
    *
+   * You should lock externally with {@link MasterWorkerInfo#lock(EnumSet, boolean)}
+   * with {@link MasterWorkerInfo.LockType#USAGE_LOCK} specified.
+   * An exclusive lock is required.
+   *
    * @param usedBytesOnTiers used bytes on each storage tier
    */
   public void updateUsedBytes(Map<String, Long> usedBytesOnTiers) {
-    mUsedBytes = 0;
-    mUsedBytesOnTiers = new HashMap<>(usedBytesOnTiers);
-    for (long t : mUsedBytesOnTiers.values()) {
-      mUsedBytes += t;
+    long usedBytes = 0;
+    mUsage.mUsedBytesOnTiers = new HashMap<>(usedBytesOnTiers);
+    for (long t : mUsage.mUsedBytesOnTiers.values()) {
+      usedBytes += t;
     }
+    mUsage.mUsedBytes = usedBytes;
   }
 
   /**
    * Sets the used space of the worker in bytes.
    *
+   * You should lock externally with {@link MasterWorkerInfo#lock(EnumSet, boolean)}
+   * with {@link MasterWorkerInfo.LockType#USAGE_LOCK} specified.
+   * An exclusive lock is required.
+   *
    * @param tierAlias alias of storage tier
    * @param usedBytesOnTier used bytes on certain storage tier
    */
   public void updateUsedBytes(String tierAlias, long usedBytesOnTier) {
-    mUsedBytes += usedBytesOnTier - mUsedBytesOnTiers.get(tierAlias);
-    mUsedBytesOnTiers.put(tierAlias, usedBytesOnTier);
+    mUsage.mUsedBytes += usedBytesOnTier - mUsage.mUsedBytesOnTiers.get(tierAlias);
+    mUsage.mUsedBytesOnTiers.put(tierAlias, usedBytesOnTier);
+  }
+
+  /**
+   * Enum for the locks in {@link MasterWorkerInfo}.
+   * Each lock type corresponds to one group of metadata in the object.
+   */
+  public enum LockType {
+    STATUS_LOCK,
+    USAGE_LOCK,
+    BLOCKS_LOCK;
+  }
+
+  /**
+   * Locks the corresponding locks on the metadata groups.
+   * The locks will be acquired in order.
+   * The locks can either be shared or exclusive.
+   * The isShared flag will apply to all the locks acquired here.
+   *
+   * @param lockTypes the locks
+   * @param isShared if false, the locking is exclusive
+   */
+  // TODO(jiacheng): refactor this to support LockResource instead of try-finally
+  public void lock(EnumSet<LockType> lockTypes, boolean isShared) {
```

Comment:
I wasn't able to achieve `lock(STATUS_LOCK  | USAGE_LOCK, true)` because java doesn't support using enum like integers (unlike Go). This seems to be the recommended way https://stackoverflow.com/questions/6282619/is-there-a-way-to-bitwise-or-enums-in-java

Compared to add an int value to each enum and bitwise with `lock(STATUS_LOCK.value | USAGE_LOCK.value)`, I prefer using `EnumSet` because it guards the argument to be enum instead of a random int. 

I guess you were thinking about sth like [this](https://github.com/Alluxio/alluxio/blob/e6b7900126ebb8d236410f35ad69b6ffdb258470/core/client/fs/src/main/java/alluxio/client/ReadType.java#L77)?

Reply:
I see. I think `EnumSet` is neat.